ggshield/config.py

```diff
@@ -23,139 +24,198 @@ def replace_in_keys(data: Dict, old_char: str, new_char: str) -> None:
-                if isinstance(getattr(self, key), list):
-                    getattr(self, key).extend(item)
-                elif isinstance(getattr(self, key), set):
-                    getattr(self, key).update(item)
-                else:
-                    setattr(self, key, item)
+    with open(path, "r") as f:
+        try:
+            data = yaml.safe_load(f) or {}
+        except Exception as e:
+            message = f"Parsing error while reading {path}:\n{str(e)}"
+            if raise_exc:
+                raise click.ClickException(message) from e
             else:
-                click.echo("Unrecognized key in config: {}".format(key))
+                click.echo(message)
+                return None
+        else:
+            return data
 
-    def load_config(self, filename: str) -> bool:
-        if not os.path.isfile(filename):
-            return False
 
-        with open(filename, "r") as f:
-            try:
-                _config = yaml.safe_load(f) or {}
-                replace_in_keys(_config, "-", "_")
-                self.update_config(**_config)
-            except Exception as e:
-                raise click.ClickException(
-                    f"Parsing error while reading {filename}:\n{str(e)}"
-                )
+def get_global_path(filename: str) -> str:
+    return os.path.join(os.path.expanduser("~"), filename)
 
-        return True
 
-    def load_configs(self, filenames: List[str]) -> None:
-        """Loads config files until one succeeds."""
-        for filename in filenames:
-            try:
-                if self.load_config(filename):
-                    return
-            except Exception as exc:
-                click.echo(str(exc))
-
-    def to_dict(self) -> Dict[str, Any]:
-        _config = {key: getattr(self, key) for key in self.get_attributes_keys()}
-        # Convert all sets into more human readable lists
-        for key in self.get_attributes_keys():
-            value = _config[key]
-            if type(value) is set:
-                _config[key] = list(value)
-        replace_in_keys(_config, "_", "-")
-        return _config
-
-    def save(self) -> bool:
+class YamlFileConfig:
+    """Helper class to define configuration object loaded from a YAML file"""
+
+    def __init__(self, **kwargs: Any) -> None:
+        raise NotImplementedError
+
+    def to_dict(self) -> Dict:
+        data = asdict(self)
+        replace_in_keys(data, old_char="_", new_char="-")
+        return data
+
+    @classmethod
+    def from_dict(cls, data: Dict) -> "YamlFileConfig":
+        replace_in_keys(data, old_char="-", new_char="_")
+        return cls(**data)
+
+    def update_config(self, data: Dict[str, Any]) -> bool:
         """
-        Save config in the first CONFIG_LOCAL file.
-        If no local config file, creates a local .gitguardian.yaml
+        Update the current config
+        raise an error if trying to set a field that does not exist
```

Comment:
This method does not do what it says in the doc: it never raises. Also, it always return True. Either the `continue` should be replaced with a `return False` or the return type should be changed to None.

Reply:
Indeed, I changed the implementation because it was breaking existing tests.

I don't understand about the return type, it always return, the `continue` can just make it leave the loop and go to `return True`.

Reply:
What I mean with the return type is that this function can never return False, so the return value is not useful. What to do depends on whether callers need to know if there was a problem:

- If they need to know, then we should replace the `continue` with a `return False`.
- If they don't need to know, then the code can be simplified by changing the return type to None (and callers need to be updated to not check the return value).

Reply:
@agateau-gg the purpose I guess (this was a pre-existing function) was being able to do `if result` I guess, but yes without a `return False` it will either return true or raise so it's not needed.

However replacing `continue` with a `return False` breaks the implementation.